### PR TITLE
Add address hierarchy widget to registration page

### PR DESCRIPTION
**Pull request by sinteacon**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

This PR integrates the address hierarchy widget to the registration page. We just create an address hierarchy component based on the layout address format and replace the existing Address Field component. The plan description about the work is on https://talk.openmrs.org/t/o3-address-hierarchy-widget-plan/36572 

## Screenshots
adH_CPT2206200933-427x338

image

*None.*
This address format is based on the layout on ethiohri

## Related Issue

*None.*

## Other

*None.*

### CODE REVIEW

**Review by samuelmale — CHANGES_REQUESTED** [review]
Thanks for the great work, looked through real quick.

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts [hunk 1]
```diff
@@ -11,6 +11,43 @@ import camelCase from 'lodash-es/camelCase';
...
+      name,
+      defaultValue: '',
+    };
+  });
+  return {
+    addressFieldValues,
+    addressValidationSchema,
+  };
+}
+export function parseAddressTemplateXmlOld(addressTemplate: string) {
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts:50]: @samuelmale we are using ethiohris address layout which is different from existing so that is why we need to create a new parser

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  const parser = new DOMParser();
+  return parser.parseFromString(xmlDockAsString, 'text/xml');
+}
+function getTagAsDocument(tagName: string, template: XMLDocument) {
+  const tmp = template.getElementsByTagName(tagName)[0];
+  return tmp ? parseString(tmp.outerHTML) : parseString('');
+}
+
+export const AddressHierarchy: React.FC = () => {
+  const nulldata = [];
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx:21]: @samuelmale can you please explain more?

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  });
+}
+export function performAdressHirarchiWithParentSearch(addressField, parentid, query) {
+  return openmrsFetch(
+    `${AH_BASE_WS_API_URL}?addressField=${addressField}&limit=20&searchString=${query}&parentUuid=${parentid}`,
+    {
+      method: 'GET',
+    },
+  );
+}
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx]: @samuelmale Thanks, I have taken the comments and I have added API layer

**Review by sinteacon — COMMENTED** [review]
@samuelmale  I have resolved the requested changes and made a new commit

File: packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.test.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import { fireEvent, render, screen, wait } from '@testing-library/react';
+import { Form, Formik } from 'formik';
+import { ComboInput } from './comboinput.component';
+
+describe.skip('Combo box input', () => {
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.test.tsx:5]: comboinput mainly uses combobox as an input so I tested Combobox labletext

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx:21]: yes, I don't need that variable I will remove it now

File: packages/esm-outpatient-app/translations/en.json [hunk 5]
```diff
@@ -89,7 +88,6 @@
   "program": "Program",
   "provider": "Provider",
   "quantity": "Quantity",
-  "recentScheduledVisits": " {count, plural, one{{count} visit} other{{count} visits}} scheduled for +/- 7 days",
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-outpatient-app/translations/en.json:92]: okay

File: packages/esm-patient-registration-app/translations/en.json [hunk 6]
```diff
@@ -4,8 +4,6 @@
   "addressHeader": "Address",
   "allFieldsRequiredText": "All fields are required unless marked optional",
   "autoGeneratedPlaceholderText": "Auto-generated",
-  "birthdayNotInTheFuture": "Birthdate cannot be in the future",
```
[the remark this answers is not in the document]
Comment (sinteacon) [reply packages/esm-patient-registration-app/translations/en.json:7]: okay

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts [hunk 7]
```diff
@@ -11,6 +11,43 @@ import camelCase from 'lodash-es/camelCase';
 import capitalize from 'lodash-es/capitalize';
 
 export function parseAddressTemplateXml(addressTemplate: string) {
+  const templateXmlDoc = new DOMParser().parseFromString(addressTemplate, 'text/xml');
+  const nameMappings = templateXmlDoc.querySelector('nameMappings'); //.querySelectorAll('property');
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts:15]: Can you remove the commented out code

File: packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (samuelmale) [reply packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.test.tsx:5]: Kindly don't skip the test suite

File: packages/esm-patient-registration-app/src/resource.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { openmrsFetch } from '@openmrs/esm-framework';
+
+const AH_BASE_WS_API_URL = '/module/addresshierarchy/ajax/getPossibleAddressHierarchyEntriesWithParents.form';
+
+export function performAdressHirarchiWithParentSearch(addressField, parentid, query) {
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/resource.ts:5]: `performAdressHirarchiWithParentSearch` -> `performAdressHierarchyWithParentSearch`

File: packages/esm-patient-registration-app/src/resource.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { openmrsFetch } from '@openmrs/esm-framework';
+
+const AH_BASE_WS_API_URL = '/module/addresshierarchy/ajax/getPossibleAddressHierarchyEntriesWithParents.form';
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/resource.ts:3]:
```javascript
const AddressHierarchyBaseURL = '/module/addresshierarchy/ajax/getPossibleAddressHierarchyEntriesWithParents.form'

```

**Review by samuelmale — COMMENTED** [review]
Almost there, just a few more minor comments..

File: packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts (hunk 7 above)
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/patient-registration/patient-registration-utils.ts:15]: fixed

File: packages/esm-patient-registration-app/src/resource.ts (hunk 8 above)
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/resource.ts:5]: fixed

File: packages/esm-patient-registration-app/src/resource.ts (hunk 9 above)
Comment (sinteacon) [reply packages/esm-patient-registration-app/src/resource.ts:3]: fixed

**Review by sinteacon — COMMENTED** [review]
requested changes have been resolved

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx [hunk 10]
```diff
@@ -1,45 +1,42 @@
...
-function getFieldValue(field: string, doc: XMLDocument) {
-  const fieldElement = doc.getElementsByTagName(field)[0];
-  return fieldElement ? fieldElement.getAttribute('value') : null;
-}
-
-export const AddressField: React.FC = () => {
-  const { addressTemplate } = useContext(ResourcesContext);
-  const [addressFields, setAddressFields] = useState([]);
+export const AddressHierarchy: React.FC = () => {
+  const [comboboxlist] = useState([]);
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx:21]: I feel like you don't need this, but if your not happy with throwing away the variable then:

`comboboxlist` --> `comboboxList`

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx [hunk 11]
```diff
@@ -1,45 +1,42 @@
...
+  const [selected, setSelected] = useState('');
+  const [addressconfig, setaddressconfig] = useState([]);
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx:23]: `const [addressConfig, setAddressConfig] = useState([]);`

File: packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx [hunk 12]
```diff
@@ -48,15 +45,27 @@ export const AddressField: React.FC = () => {
...
+      <div
+        style={{
+          width: '50%',
+          paddingBottom: '5%',
+        }}>
+        {addressconfig.map((attributes) => (
+          <ComboInput
+            name={attributes.name}
+            labeltext={attributes.labelText}
+            items={comboboxlist}
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/patient-registration/field/address/address-hierarchy.component.tsx:63]: Why don't you just pass `[]` instead?

File: packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.component.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+interface InputProps extends ComboBoxProps {
+  name: string;
+  labeltext: string;
+  setSelectedValue: any;
+  selected: string;
+}
+
+export const ComboInput: React.FC<InputProps> = ({ name, labeltext, setSelectedValue, selected }) => {
+  const [field, Meta, helpers] = useField(name);
+  const [comboboxlist, setcomboboxlist] = useState([]);
```
Comment (samuelmale) [inline packages/esm-patient-registration-app/src/patient-registration/input/comboinput/comboinput.component.tsx:15]: I would rename this to:

`const [items, setItems] = useState([]);`

**Review by samuelmale — APPROVED** [review]
Looks great overall, just a few more minor comments.

**Comment by samuelmale** [comment]
@sinteacon any updates on this?

**Comment by vasharma05** [comment]
Hi @sinteacon!
Please have a look at the comments above and do update about the status of the PR.
Thanks!

**Comment by sinteacon** [comment]
> Hi @sinteacon! Please have a look at the comments above and do an update about the status of the PR. Thanks!

I have resolved the mentioned comments thanks !

**Comment by makombe** [comment]
@sinteacon Awesome work on the address hierarchy. The feature will be central for us as we migrate KenyaEMR to OpenMRS 3.0

File: packages/esm-patient-registration-app/src/patient-registration/field/dob/dob.component.tsx [hunk 14]
```diff
@@ -65,22 +93,14 @@ export const DobField: React.FC = () => {
...
-              {...birthdate}
-              placeholder={placeHolder}
-              labelText={t('dateOfBirthLabelText', 'Date of Birth')}
-              invalid={!!(birthdateMeta.touched && birthdateMeta.error)}
-              invalidText={birthdateMeta.error && t(birthdateMeta.error)}
-              value={format(birthdate.value)}
-            />
-          </DatePicker>
+          <Provider locale="am-AM-u-ca-ethiopic" colorScheme="light" theme={defaultTheme}>
+            <DatePicker
```
Comment (donaldkibet) [inline packages/esm-patient-registration-app/src/patient-registration/field/dob/dob.component.tsx:97]: Any reasons for removing the `DatePickerInput` component for the DatePicker

File: packages/esm-patient-search-app/src/patient-search-result/patient-search-result.component.tsx [hunk 15]
```diff
@@ -27,7 +46,7 @@ const PatientSearchResults: React.FC<PatientSearchResultsProps> = ({ patients, h
           },
         ],
         gender: patient.person.gender,
-        birthDate: patient.person.birthdate,
+        birthDate: gregToEth(patient.person.birthdate),
```
Comment (donaldkibet) [inline packages/esm-patient-search-app/src/patient-search-result/patient-search-result.component.tsx:49]: This should be configurable to be able to handle both the Ethiopian date and the Gregorian date formats

**Review by donaldkibet — CHANGES_REQUESTED** [review]
Thanks @sinteacon for this work, a few comments

File: package.json [hunk 16]
```diff
@@ -1,5 +1,6 @@
 {
-  "name": "@openmrs/esm-patient-management",
+  "name": "@icap-ethiopia/esm-patient-management",
```
Comment (pirupius) [inline package.json:2]: @sinteacon it seems changes from ur fork sneaked into this PR. Please review the files and revert these changes

File: .npmrc [hunk 17]
```diff
@@ -0,0 +1 @@
+@icap-ethiopia:registry=https://npm.pkg.github.com
```
Comment (pirupius) [inline .npmrc:1]: same comment about ur custom fork changes

**Review by pirupius — CHANGES_REQUESTED** [review]
@sinteacon please clean this branch and remove any forked residue u added. Also limit the changes to only the files that are related to address heirachy

File: package.json (hunk 16 above)
Comment (sinteacon) [reply package.json]: sorry for the messed up, I just reverted the changes.

File: .npmrc (hunk 17 above)
Comment (sinteacon) [reply .npmrc]: resolved

**Comment by sinteacon** [comment]
@pirupius conflict resolved

**Comment by pirupius** [comment]
@donaldkibet can you do another pass at this?

**Review by pirupius — APPROVED** [review]
